src/lib/Microsoft.Fx.Portability.Reports.DGML/ReferenceNode.cs

```diff
@@ -66,65 +66,65 @@ public double GetPortabilityIndexForReferences(int target)
             if (Nodes.Count == 0)
                 return 1;
 
+            // reset all the nodes in the reference graph as not be searched.
+            ResetReferenceSearchGraph();
+
             // sum up the number of calls to available APIs and the ones for not available APIs for references.
-            if (!TryGetAPICountFromReferences(target, out int availableApis, out int unavailableApis))
-            {
-                // Cycle detected
-                return 1;
-            }
-            else
-            {
-                // remove the calls from the current node.
-                availableApis -= UsageData[target].GetAvailableAPICalls();
-                unavailableApis -= UsageData[target].GetUnavailableAPICalls();
+            GetAPICountFromReferences(target, out int availableApis, out int unavailableApis);
 
-                // prevent Div/0
-                if (availableApis == 0 && unavailableApis == 0)
-                    return 0;
+            // prevent Div/0. When availableApis is 0 and unavailableApis is 0, it likely means that it's references are all not resulted assemblies, return 0.
+            if (availableApis == 0 && unavailableApis == 0)
+                return 0;
 
-                return availableApis / ((double)availableApis + unavailableApis);
-            }
+            return availableApis / ((double)availableApis + unavailableApis);
         }
 
-        public bool TryGetAPICountFromReferences(int target, out int availAPIs, out int unavailAPIs)
+        private void GetAPICountFromReferences(int target, out int availAPIs, out int unavailAPIs)
         {
-            availAPIs = UsageData[target].GetAvailableAPICalls();
-            unavailAPIs = UsageData[target].GetUnavailableAPICalls();
-
-            // We are going to use a flag on the object to detect if we have a reference cycle while computing the APIs for the references.
-            if (_searchInGraph == true)
-            {
-                // Cycle!!!
-                _searchInGraph = false; // Reset this flag
-                return false;
-            }
-            else
-            {
-                _searchInGraph = true;
-            }
+            availAPIs = 0;
+            unavailAPIs = 0;
 
             foreach (var item in Nodes)
             {
-                if (item.UsageData == null)
+                // We are going to use the flag on the object to detect if we have run into this node in the reference graph while computing the APIs for the references,
+                // because we want to only count a reference node once if it is referenced in the graph more than once.
+                if (item._searchInGraph == true)
                 {
-                    // skip the Node with no UsageData
                     continue;
                 }
-
-                if (!item.TryGetAPICountFromReferences(target, out int refCountAvail, out int refCountUnavail))
+                else
                 {
```

Comment:
Super NIT: I don't like the "else" here, because technically, the rest of the foreach loop is the else. This is more personal code philosophy though, so it's not a big deal.

Reply:
It's gone in updated code.